### DISCUSSION TITLE
When does always get triggered?

### DISCUSSION THREAD

**Discussion by tinitus-git** [statelyai/xstate#3056] [General]
I am wondering when always is triggered and why it is not triggered in the following sceanrio.

If I have a machine with the following definition:

`
import { createMachine } from "xstate";

const genericMachine = createMachine(
  {
    id: "FsmMachine",
    context: {
      StateVar: "",
    },
    initial: "State1",
    states: {
      State1: {
        always: {
          target: "State2",
          cond: "StateVarCondition",
        },
        on: {
          AssignValue: {
            actions: "AssignStateVar",
          },
        },
      },
      State2: {
        type: "final",
      },
    },
  },
  {
    actions: {
      AssignStateVar: (context, event) => {
        context.StateVar = event.data;
      }
    },
    guards: {
      StateVarCondition: (context) => {
        return context.StateVar !== "";
      },
    },
  }
);
 `

As you can see, the machine consists of two states. The first state has an event driven transition defined. When this event is triggered, a variable in the context data is set. The first state also has an always transition that has a condition that looks to see if the context variable has a value other than empty string.

If I send the following event (either through visualizer of in code):

`
{
  "type: "AssignValue",
  "data": "SomeValue"
}
`

I see the context variable value has changed, but the machine is still at State1. I would have thought the transition to assign the value would trigger the always transition on re-entry to the state.

I have a number of workaround for this (including sending a blank event that triggers the alway transition, but I am wondering why a workaround such as this is needed. Is this works as designed, or is this bug?

**Comment by tinitus-git** [comment]
Closing this because it is a duplicate of another issue I created.

**Comment by Andarist** [comment]
`always` should get triggered after every other transition (including always transitions). The rough pseudocode of the algorithm that is close to what we are using can be found [here](https://www.w3.org/TR/scxml/#mainEventLoop). You can think of `selectEventlessTransitions` as selecting always transitions.

**Reply by tinitus-git** [reply]
So if I understand properly, the always transition should get triggered after the AssignValue event which triggers an event back to State1? If so, then I would expect the AssignValue event triggering the transition back to State1 and then the always transition should trigger and the machine should end up in State2. Is this then a bug? It seems to me that this could due to one of the following:
1. The always actually is not being triggered. OR
2. The context variable does not have the proper value when the always transition is triggered and therefore the condition is not evaluating to true.

**Reply by tinitus-git** [reply]
Doing a little bit of debugging, I *think* the problem I am seeing is due to the context variable not having a value at the time of the always condition being evaluated. Is there a way to "sync" the context value? Thanks.

**Reply by tinitus-git** [reply]
Sorry I think I spoke a bit too soon. It appears the always condition is triggered prior to the assignment action.
